src/webgl/loading.js

```diff
@@ -437,14 +419,14 @@ function parseObj(model, lines, materials= {}) {
           ) {
             model.faces.push(face);
             //same material for all vertices in a particular face
-            const materialDiffuseColor =
-             materials[currentMaterial].diffuseColor;
-            for (let i=0 ;i<face.length;i++) {
-              model.vertexColors.push([
-                materialDiffuseColor[0],
-                materialDiffuseColor[1],
-                materialDiffuseColor[2]
-              ]);
+            if (currentMaterial && materials[currentMaterial]) {
```

Comment:
Right now this will break if we have some vertices that don't have a material but others that do (since it means the earlier vertices won't have pushed colors to the array.) Our data needs it to be all or nothing.

I don't think we need to actually support that case, but maybe we can try to detect it and throw an error if we encounter it? i.e. if we've added any vertices with no colors, and then find ourselves trying to add a vertex *with* color later on, throw an error letting people know they need colors for all their vertices or none.

Reply:
Thinking about it, there can be multiple ways of ensuring colors or no colors for all vertices, one being the first vertex added would dictate if all the further vertices to be added must have a color or not. i.e. if the first vertex has a color, all the vertices of the model must be colored and vice versa thus, all or nothing for colors of vertices.
Using a flag `colorRequirementSet` for this. Haven't thoroughly checked though if this too would break.

Reply:
```
function parseObj(model, lines, materials= {}) {
  // OBJ allows a face to specify an index for a vertex (in the above example),
  // but it also allows you to specify a custom combination of vertex, UV
  // coordinate, and vertex normal. So, "3/4/3" would mean, "use vertex 3 with
  // UV coordinate 4 and vertex normal 3". In WebGL, every vertex with different
  // parameters must be a different vertex, so loadedVerts is used to
  // temporarily store the parsed vertices, normals, etc., and indexedVerts is
  // used to map a specific combination (keyed on, for example, the string
  // "3/4/3"), to the actual index of the newly created vertex in the final
  // object.
  const loadedVerts = {
    v: [],
    vt: [],
    vn: []
  };


  // Map from source index → Map of material → destination index
  const usedVerts = {}; // Track colored vertices
  let currentMaterial = null;
  let uniqueVertices = 0; // Set to track unique vertices
  let coloredUniqueVertexCount = 0;
  let hasColoredVertices = false;
  for (let line = 0; line < lines.length; ++line) {
    // Each line is a separate object (vertex, face, vertex normal, etc)
    // For each line, split it into tokens on whitespace. The first token
    // describes the type.
    const tokens = lines[line].trim().split(/\b\s+/);

    if (tokens.length > 0) {
      if (tokens[0] === 'usemtl') {
        // Switch to a new material
        currentMaterial = tokens[1];
      }else if (tokens[0] === 'v' || tokens[0] === 'vn') {
        // Check if this line describes a vertex or vertex normal.
        // It will have three numeric parameters.
        const vertex = new p5.Vector(
          parseFloat(tokens[1]),
          parseFloat(tokens[2]),
          parseFloat(tokens[3])
        );
        if (tokens[0]==='v'){
          uniqueVertices++;
        }
        loadedVerts[tokens[0]].push(vertex);
      } else if (tokens[0] === 'vt') {
        // Check if this line describes a texture coordinate.
        // It will have two numeric parameters U and V (W is omitted).
        // Because of WebGL texture coordinates rendering behaviour, the V
        // coordinate is inversed.
        const texVertex = [parseFloat(tokens[1]), 1 - parseFloat(tokens[2])];
        loadedVerts[tokens[0]].push(texVertex);
      } else if (tokens[0] === 'f') {
        // Check if this line describes a face.
        // OBJ faces can have more than three points. Triangulate points.
        for (let tri = 3; tri < tokens.length; ++tri) {
          const face = [];
          const vertexTokens = [1, tri - 1, tri];

          for (let tokenInd = 0; tokenInd < vertexTokens.length; ++tokenInd) {
            // Now, convert the given token into an index
            const vertString = tokens[vertexTokens[tokenInd]];
            let vertParts=vertString.split('/');

            // TODO: Faces can technically use negative numbers to refer to the
            // previous nth vertex. I haven't seen this used in practice, but
            // it might be good to implement this in the future.

            for (let i = 0; i < vertParts.length; i++) {
              vertParts[i] = parseInt(vertParts[i]) - 1;
            }

            if (!usedVerts[vertParts[0]]) {
              usedVerts[vertParts[0]] = {};
            }

            if (usedVerts[vertParts[0]][currentMaterial] === undefined) {
              const vertIndex = model.vertices.length;
              model.vertices.push(loadedVerts.v[vertParts[0]].copy());
              model.uvs.push(loadedVerts.vt[vertParts[1]] ?
                loadedVerts.vt[vertParts[1]].slice() : [0, 0]);
              model.vertexNormals.push(loadedVerts.vn[vertParts[2]] ?
                loadedVerts.vn[vertParts[2]].copy() : new p5.Vector());

              usedVerts[vertParts[0]][currentMaterial] = vertIndex;
              face.push(vertIndex);
              //every unique vertex would have its current material undefined once
              coloredUniqueVertexCount++;
            } else {
              face.push(usedVerts[vertParts[0]][currentMaterial]);
            }
          }

          if (
            face[0] !== face[1] &&
            face[0] !== face[2] &&
            face[1] !== face[2]
          ) {
            model.faces.push(face);
            //same material for all vertices in a particular face
            if (currentMaterial
              && materials[currentMaterial]
              && materials[currentMaterial].diffuseColor) {
              const materialDiffuseColor =
              materials[currentMaterial].diffuseColor;
              hasColoredVertices = true;
              for (let i = 0; i < face.length; i++) {
                model.vertexColors.push(materialDiffuseColor[0]);
                model.vertexColors.push(materialDiffuseColor[1]);
                model.vertexColors.push(materialDiffuseColor[2]);
              }
            }
          }
        }
      }
    }
  }
  // If the model doesn't have normals, compute the normals
  if (model.vertexNormals.length === 0) {
    model.computeNormals();
  }
  if (hasColoredVertices) {
    // If any vertex has been colored, check if all have been
    if (coloredUniqueVertexCount !== uniqueVertices) {
      throw new Error('All vertices must be colored if any is colored, but found a mix.');
    }
  }
  return model;
}
```

@davepagurek Would something like this make sense for an all-or-nothing, i.e. a fully colored model or a colorless model. Assuming every unique vertex should have a color or no color for all vertices at all?